File: 202101969/1717.c
# #include <stdio.h>
# #include <stdbool.h>
int parent[1000000];

int Find(int x){
    if (x == parent[x]){
        return x;
    }
    else{
        return parent[x] = Find(parent[x]);
    }
}

void Union(int x, int y){
    x = Find(x);
    y = Find(y);

    if (x==y){
        return;
    }
    if(x < y){
        parent[y] = x;
    }
    else{
        parent[x] = y;
    }
}

int main() {
    int n, m;
    scanf("%d %d", &n, &m);
    for(int i = 0; i<=n; i++){
        parent[i] = i;
    }

    for(int i=0; i<m; i++){
        int a,b,c;
        scanf("%d %d %d",&a,&b,&c);
        if (a ==0){
            Union(b,c);
        }
        else{
            int B,C;
            B = Find(b);
            C = Find(c);
            if (B == C){
                printf("YES \n");
            }
            else{
                printf("NO \n");
            }
        }
    }
    return 0;
}

File: 202102680/1717.py
import sys
n, m = map(int,input().split())
sys.setrecursionlimit(100000) # 재귀 깊이 제한 늘리기
node = [i for i in range(n+1)]

def find(x): # node[x] != x 의 의미는 다른 어딘 가에 종속되어있다는 것 => 중속된 것을 재귀로 찾아냄
    if node[x] != x:
        node[x] = find(node[x])
    return node[x]

def union(a, b):
    a = find(a)
    b = find(b)

    if a < b: # 작은 거 기준으로 종속
        node[b] = a
    else:
        node[a] = b

for _ in range(m):
    num, a, b = map(int,input().split())

    if a == b and not num:
        continue
    if num: # 두 원소가 같은 집합에 포함 되어 있는 지 확인
        if find(a) == find(b):
            print("YES")
        else:
            print("NO")
    else: # num 이 0인 경우 => 합집합 연산
        union(a,b)
